### PR TITLE
Issue #10138: Finite Temporal Helpers

### LINKED ISSUE #10138: duckdb_from_timestamp throws for infinity and -infinity
### What happens?

When trying to read `infinity` and `-infinity` timestamps with C Api, `duckdb_from_timestamp` throws an exception

### To Reproduce

```c
    duckdb_database db;
    duckdb_connection con;
    duckdb_state state;
    duckdb_result result;

    if (duckdb_open(NULL, &db) == DuckDBError) {
        // handle error
    }
    if (duckdb_connect(db, &con) == DuckDBError) {
        // handle error
    }

    state = duckdb_query(con, "SELECT timestamp_array FROM test_all_types()", &result);
    duckdb_data_chunk chunk = duckdb_result_get_chunk(result, 0);
    idx_t row_count = duckdb_data_chunk_get_size(chunk);
    duckdb_vector vector = duckdb_data_chunk_get_vector(chunk, 0);
    duckdb_vector childVector= duckdb_list_vector_get_child(vector);

    duckdb_timestamp* timeData = (duckdb_timestamp *)duckdb_vector_get_data(childVector);

    //Works fine for the first element which is 1970, 1, 1 but throws for second and third
    duckdb_timestamp_struct timestampStruct = duckdb_from_timestamp(timeData[0]);
    timestampStruct = duckdb_from_timestamp(timeData[1]);

// cleanup
    duckdb_destroy_data_chunk(&chunk);
    duckdb_destroy_result(&result);
    duckdb_disconnect(&con);
    duckdb_close(&db);
```

image
image

### OS:

Windows 11

### DuckDB Version:

0.9.2

### DuckDB Client:

C

### Full Name:

Giorgi Dalakishvili

### Affiliation:

Space International

### Have you tried this on the latest `main` branch?

I have tested with a release build (and could not test with a main build)

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by hawkfish**
Add two C APIs for detecting whether a date or a timestamp is finite.

fixes: #10138
fixes: duckdblabs/duckdb-internal#1005

### CODE REVIEW

**Comment by hawkfish** [comment]
The theme song for this issue is of course The Who's _The Song is Over_

> I'll sing my heart out to the infinite C

File: src/include/duckdb.h [hunk 1]
```diff
@@ -887,6 +887,14 @@ Re-compose a `duckdb_date` from year, month and date (`duckdb_date_struct`).
 */
 DUCKDB_API duckdb_date duckdb_to_date(duckdb_date_struct date);
 
+/*!
+Test a `duckdb_date` to see if it is a finite value.
+
+* date: The date object, as obtained from a `DUCKDB_TYPE_DATE` column.
+* returns: True if the date is finite, false if it is ±infinity.
+*/
+DUCKDB_API bool duckdb_is_finite_date(duckdb_date_struct date);
```
Comment (Mytherin) [inline src/include/duckdb.h:896]: This should be `duckdb_date` not `duckdb_date_struct`

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM - one comment. Could you also add some tests (e.g. in `test/api/capi/test_capi_data_chunk.cpp` or `test/api/capi/test_capi_complex_types.cpp`)?

**Comment by hawkfish** [comment]
> Thanks for the PR! LGTM - one comment. Could you also add some tests (e.g. in `test/api/capi/test_capi_data_chunk.cpp` or `test/api/capi/test_capi_complex_types.cpp`)?

Sorry I wasn't sure where they should go and I couldn't find any for the other functions.

**Comment by Mytherin** [comment]
In this case the `duckdb_timestamp` and `duckdb_date` values are generally only obtained as part of the (newer) data chunk scanning interface (which mirrors the C++ interface and essentially provides direct access to DuckDB's `Vector` class through the C API).

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**